### ISSUE TITLE
Migrate Memacs to Python 3

### ISSUE THREAD

**Issue by novoid** [novoid/Memacs#52]
This is not complicated due to `2to3` which does most of the work.

However, many files have to be changed and many tests to be run.

**Comment by DerBeutlin** [comment]
I started working on this issue [here](https://github.com/DerBeutlin/Memacs/tree/port_to_python3) but I have a question:
- One crucial difference between python2 and python3 is the order iteratively added items are stored in dictionaries. This results in many unittests failing because the order of the properties is wrong. Originally I wanted to keep the order as stated in the unittests but this is sometimes very ugly when reading data from file and adding the properties after each line.  Now the question is: is the order of the properties important? Because if not one could either rewrite the unittest such that they just check that every property exists and is correct but would ignore the order. This would have the benefit that the functionality could stay the same for python3 and python2.

**Comment by novoid** [comment]
Bummer.
I was not aware of that. My other projects I migrated did not use that dependency.
IMO: the order of the properties does not have any information contained implicitly. So if the IDs have to be changed once and then the Python 3 compatibility is done, I think we're fine.

The only downside is that re-processing old data with the Python 3 Memacs modules result in different IDs which may cause broken links :-(

AFAIR, I personally did not link Memacs-generated items yet but others might have. So this is a breaking change which must be well documented. Keep me updated and I do the documenting stuff.

Thanks for your work - very much appreciated!

**Comment by DerBeutlin** [comment]
If I'm not mistaken the order in the dictionary is random but reproducible in python2, differently random and also reproducible in python3.1-python3.5 and starting with python3.6 it stays in the order it was inserted (so reproducible and not random). 
IMO opinion there are two options:
1) don't enforce any particular order of the properties, test only for existence, then everybody using memacs with python2 (given that compatibility will remain) keeps the same entries and the same ID's, but as soon as they cross the line to python3 or to python3.6 everything changes.
2) enforce insertion order of the properties using OrderedDicts. Then everybody would have the same entries no matter what python version they're using, but also everybody would have a potential broken links.

I have a slight preference for option 2 but if python2.7 support is kept, then with option1 everybody could decide whether or not he/she can risk an upgrade of the python version

**Comment by nico202** [comment]
What about https://docs.python.org/3/library/collections.html#collections.OrderedDict ?

**Comment by DerBeutlin** [comment]
@nico202 this is exactly what I mentioned in 2). it would break the current order of the properties and therefore the ID's but would otherwise be a nice solution.

**Comment by novoid** [comment]
Hi,
I think we mixed up two things here (or at least I did):

- unit tests and changed IDs whose comparison will fail
- potential data change at the current users' installation that might break existing links

For the first one, the remarks of you both are valid and I tend to follow the least effort here: don't enforce order, ignore IDs in unit test comparison.

The second issue still prevails and I don't think that there is a good solution which will not result in changed IDs for existing Memacs files generated by Python 2 Memacs.

**Comment by DerBeutlin** [comment]
Yes, I think the first issue can be handled easily no matter which of the two approaches we're using. The second one boils down to break now ID's from python2 and be consistent afterwards no matter what happens or support the old ID's  for all users of python2 and introduce the break only when they change python versions (but then it will break again when they're moving from python3.5 to python3.6)

**Comment by novoid** [comment]
First issue: the one implementing/fixing/migrating may decide. (I go for minimal effort as explained above).

Second issue: document well and inform the users - I can handle that.

**Comment by novoid** [comment]
I personally do favor Python 3 such that Python 2 backward compatibility is not worth much effort (any more). Python 3 is the way to go since a couple of years.

As far as I know, all this explicit encoding to UTF-8/unicode is obsolete with Python 3.

My personal workflow for all of my Python projects I migrated from 2 to 3:

1. converting Python 2 to 3: http://python3porting.com/
    - `2to3 -w myoldscript.py` -> writing
2. adapt shebang "python(2)"→"python3"
    - `#!/usr/bin/env python3`
3. check for unicode checks
    - `assert(type(tag) == unicode)` - there might be redundant checks after 2to3: `assert(x=str or x=str)`

With this simple process, I migrated approximately 15 small(er) Python tools and https://karl-voit.at/tags/lazyblorg/

**Comment by novoid** [comment]
FYI: I just ran all unit tests for all modules so that their tests worked. One exception: https://github.com/novoid/Memacs/commit/ebd6861f345b0bf7502996c0b62de9ec9c08cf5a

The script I am using:

```
#!/bin/sh

PYTHONPATH="~/src/memacs/memacs:" python3 -m pytest ./memacs/lib/tests/orgwriter_test.py && \
PYTHONPATH="~/src/memacs/memacs:" python3 -m pytest ./memacs/lib/tests/orgproperty_test.py && \
PYTHONPATH="~/src/memacs/memacs:" python3 -m pytest ./memacs/lib/tests/reader_test.py && \
PYTHONPATH="~/src/memacs/memacs:" python3 -m pytest ./memacs/lib/tests/argparser_test.py && \
PYTHONPATH="~/src/memacs/memacs:" python3 -m pytest ./memacs/lib/tests/mailparser_test.py && \
PYTHONPATH="~/src/memacs/memacs:" python3 -m pytest ./memacs/lib/tests/orgformat_test.py && \
PYTHONPATH="~/src/memacs/memacs:" python3 -m pytest ./memacs/tests/sms_test.py && \
PYTHONPATH="~/src/memacs/memacs:" python3 -m pytest ./memacs/tests/phonecalls_test.py && \
PYTHONPATH="~/src/memacs/memacs:" python3 -m pytest ./memacs/tests/git_test.py && \
PYTHONPATH="~/src/memacs/memacs:" python3 -m pytest ./memacs/tests/filenametimestamps_test.py && \
PYTHONPATH="~/src/memacs/memacs:" python3 -m pytest ./memacs/tests/csv_test.py && \
PYTHONPATH="~/src/memacs/memacs:" python3 -m pytest ./memacs/tests/battery_test.py && \
PYTHONPATH="~/src/memacs/memacs:" python3 -m pytest ./memacs/tests/simplephonelogs_test.py && \
PYTHONPATH="~/src/memacs/memacs:" python3 -m pytest ./memacs/tests/example_test.py && \
PYTHONPATH="~/src/memacs/memacs:" python3 -m pytest ./memacs/tests/whatsapp_test.py && \
PYTHONPATH="~/src/memacs/memacs:" python3 -m pytest ./memacs/tests/photos_test.py && \
PYTHONPATH="~/src/memacs/memacs:" python3 -m pytest ./memacs/tests/arbtt_test.py && \
PYTHONPATH="~/src/memacs/memacs:" python3 -m pytest ./memacs/tests/svn_test.py && \
PYTHONPATH="~/src/memacs/memacs:" python3 -m pytest ./memacs/tests/orgformat_test.py && \
PYTHONPATH="~/src/memacs/memacs:" python3 -m pytest ./memacs/tests/ical_test.py && \
PYTHONPATH="~/src/memacs/memacs:" python3 -m pytest ./memacs/tests/rss_test.py && \
PYTHONPATH="~/src/memacs/memacs:" python3 -m pytest ./memacs/tests/gpx_test.py && \
echo "    =========================>  ALL TESTS RAN WITHOUT ERROR  <========================="

#end
```

Should this script be checked in? What do you think?

What are the remaining tasks that this issue can be closed? (Are we done?)

Thank you **very much** for all your effort on this issue!

**Comment by DerBeutlin** [comment]
I use the test runner included in spacemacs which uses [pytest.el](https://github.com/ionrock/pytest-el) which work pretty well.

Nonetheless it would be nice to have an easy way to start all the tests however there are two problems with the script
1) Global path to memacs is different for everybody
2) pytest should be able to find all tests by itself so it should not be necessary to list all the tests in the script, does it? 

So my suggestion would just be to recommend to the user to set an the variable 
```
export MEMACSPATH="~/src/memacs/memacs"
```
 in their `.bashrc` 
and then call
```
PYTHONPATH=$MEMACSPATH python3 -m pytest $MEMACSPATH
```
 which calls all tests from everywhere in the system. That way if one adds a module one does not have to add it to the script.

**Comment by novoid** [comment]
> With this simple process, I migrated approximately 15 small(er) Python tools and https://karl-voit.at/tags/lazyblorg/

OMG: I just recognized that I never migrated my lazyblorg to Python3 :-O
... preparing for just another migration right now ... :-(
